Add an implementation of the GProximal interface for TensorFlow model(s)

tkoskela opened the pull request:
Closes #299 
Closes #331 

TODO list
- [x] Add calls to pre and post process the image data into the right format for the model
- [x] Add tf_g_proximal.h to CMakeLists.txt
- [x] Check that it compiles
- [x] Create a test for it -- inpainting?
- [x] Instantiate cppflow::model in constructor
- [x] Clean up image conversions between t_Vector and cppflow::tensor
- [x] Figure out why solution does not converge
- [x] #331

tkoskela committed:
Add draft for TF implementation + some TODOs

SJaffa commented:
CppFlow docs on finding the model call signature is in the note at the end of this section: https://serizba.github.io/cppflow/quickstart.html#complex-model-call-multi-input-output-models

tkoskela committed:
Fix variable names in proximal

tkoskela committed:
Merge branch 'tk/fix-conan-macos-issue' into tk/tf-proximal-implementation

tkoskela committed:
Add header to CMakeLists for compiling

tkoskela committed:
Do post and preprocessing of image data

tkoskela committed:
Add test using TF proximal

tkoskela committed:
Add support for non-square images

tkoskela committed:
Merge branch 'tk/g_proximal_args' into tk/tf-proximal-implementation

tkoskela committed:
Merge branch 'development' into tk/tf-proximal-implementation

tkoskela committed:
Bug fixes. WIP

tkoskela committed:
Fix unit weights

tkoskela committed:
Test runs but solution does not converge.

Issues with conflicting data types should be resolved.
There is an error in the model call if I instantiate the model in the
constructor. I left some commented out code for debugging this,
currently the model is read on every model call which is extremely
slow and definitely not how we want to do it.
Because the data type of the image is t_Vector (instead of Image) the
functions in cppflowutils did not work directly. I overloaded the
preprocessing function but did not yet do anything with the
postprocessing function.

tkoskela committed:
Create model object in constructor

tkoskela committed:
Make the model const, we are not changing it

tkoskela committed:
Update cppflow submodule to use const operator()

tkoskela committed:
Merge branch 'development' into tk/tf-proximal-implementation

tkoskela committed:
Clean up post processing, add notes on types

tkoskela committed:
add template keyword to fix apple-clang error

tkoskela committed:
Merge branch 'development' into tk/tf-proximal-implementation

tkoskela committed:
Merge branch 'development' into tk/tf-proximal-implementation

tkoskela committed:
Merge branch 'development' into tk/tf-proximal-implementation

tkoskela commented:
In this PR, we are still always using the FISTA algorithm, do you think that is OK or do we still need to do #331? We would need to parametrize how `t` is updated in https://github.com/astro-informatics/sopt/blob/8b49e78cbcb824db084392f02c4f2dd6b5cf328d/cpp/sopt/forward_backward.h#L255

tkoskela committed:
Add bool switch for FISTA algorithm. True by default, set by setter.

tkoskela committed:
Move calculations out of arguments to g_proximal

tkoskela committed:
Set conan version to avoid getting 2.0

SJaffa commented on cpp/sopt/tf_g_proximal.h:
```diff
@@ -0,0 +1,122 @@
+#include "sopt/types.h"
+#include "sopt/g_proximal.h"
+
+#include <cppflow/cppflow.h>
```
```suggestion
#include "cppflow/cppflow.h"
```

SJaffa commented on cpp/tests/tf_inpainting.cc:
```diff
@@ -0,0 +1,86 @@
+#include <ctime>
+#include <catch.hpp>
+
+#include <sopt/imaging_forward_backward.h>
```
```suggestion
#include "sopt/imaging_forward_backward.h"
```

SJaffa commented on cpp/tests/tf_inpainting.cc:
```diff
@@ -0,0 +1,86 @@
+    .beta(beta)    // stepsize
+    .sigma(sigma)  // sigma
+    .gamma(gamma)  // regularisation paramater
+    .fista(false)  // switch for FISTA algorithm, don't use with learned TF model
```
```suggestion
    .fista(false)  // switch to use FISTA algorithm in Forward Backward algorithm, should be false if using learned TF model
```

tkoskela committed:
Improve comments for doxygen

SJaffa commented on cpp/sopt/tf_g_proximal.h:
```diff
@@ -0,0 +1,122 @@
+      square_image_(true),
+      Psi_(linear_transform_identity<Scalar>())
+  {
+  //   for ( auto str : model_.get_operations() ) {
```
Should we remove this print-debugging or keep it for later use?

tkoskela committed:
Use common convention for includes

tkoskela committed:
Clean up commented out code

tkoskela committed:
Update cpp/tests/tf_inpainting.cc

Co-authored-by: S Jaffa <SJaffa@users.noreply.github.com>
